activation/activation.go

```diff
@@ -244,34 +249,47 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
 func (b *Builder) StartPost(rewardAddress address.Address, dataDir string, space uint64) error {
 	b.log.Info("Starting post, reward address: %x", rewardAddress)
 	b.SetCoinbaseAccount(rewardAddress)
-	b.postInitLock.Lock()
-	if b.initStatus == Done {
-		return nil
+
+	initStatus := atomic.LoadInt32(&b.initStatus)
+
+	if initStatus == InitDone {
+		return fmt.Errorf("already initialized")
 	}
-	if b.initStatus == InProgress {
-		b.postInitLock.Unlock()
-		return fmt.Errorf("attempted to start post when post already initiated")
+	if initStatus == InitInProgress {
+		return fmt.Errorf("already started")
 	}
-	b.initStatus = InProgress
+
+	atomic.StoreInt32(&b.initStatus, InitInProgress)
+
 	errChan := make(chan error, 1)
 	go func() {
-		_, err := b.nipstBuilder.InitializePost(dataDir, space) // TODO: add proof to first ATX
-		b.postInitLock.Lock()
-		if err != nil {
-			b.initStatus = Idle
-			b.postInitLock.Unlock()
-			b.log.Error(err.Error())
-			errChan <- err
-			return
+		var err error
+		b.postProver.SetParams(dataDir, space)
+		if b.postProver.IsInitialized() {
+			b.commitment, err = b.postProver.Execute(shared.ZeroChallenge)
+			if err != nil {
+				errChan <- err
+				atomic.StoreInt32(&b.initStatus, InitIdle)
+				b.log.Error("PoST execution failure: %v", err)
+				return
+			}
+		} else {
+			b.commitment, err = b.postProver.Initialize()
+			if err != nil {
+				errChan <- err
+				atomic.StoreInt32(&b.initStatus, InitIdle)
+				b.log.Error("PoST initialization failure: %v", err)
+				return
+			}
 		}
-		b.initStatus = Done
-		b.postInitLock.Unlock()
+
+		b.log.Info("PoST initialization completed")
+		atomic.StoreInt32(&b.initStatus, InitDone)
 	}()
-	b.postInitLock.Unlock()
+
 	// todo: refactor initializePost to be concurrent, returning error in a blocking manner
-	t := time.NewTimer(b.errorTimeoutMs * time.Millisecond)
 	select {
-	case <-t.C:
+	case <-time.After(100 * time.Millisecond):
```

Comment:
move out to constant/config

Reply:
I removed it from the atx builder on purpose, so we won't get used to it. The timeout solution is bad and should be removed.